mm2src/coins/eth/eth_swap_v2/eth_maker_swap_v2.rs

```diff
@@ -0,0 +1,490 @@
+use super::{validate_from_to_and_status, validate_payment_args, EthPaymentType, PaymentMethod, PrepareTxDataError,
+            ZERO_VALUE};
+use crate::coin_errors::{ValidatePaymentError, ValidatePaymentResult};
+use crate::eth::{decode_contract_call, get_function_input_data, wei_from_big_decimal, EthCoin, EthCoinType,
+                 MakerPaymentStateV2, SignedEthTx, MAKER_SWAP_V2};
+use crate::{ParseCoinAssocTypes, RefundMakerPaymentSecretArgs, RefundMakerPaymentTimelockArgs, SendMakerPaymentArgs,
+            SpendMakerPaymentArgs, SwapTxTypeWithSecretHash, TransactionErr, ValidateMakerPaymentArgs};
+use ethabi::{Function, Token};
+use ethcore_transaction::Action;
+use ethereum_types::{Address, Public, U256};
+use ethkey::public_to_address;
+use futures::compat::Future01CompatExt;
+use mm2_err_handle::mm_error::MmError;
+use mm2_err_handle::prelude::MapToMmResult;
+use std::convert::TryInto;
+use web3::types::TransactionId;
+
+const ETH_MAKER_PAYMENT: &str = "ethMakerPayment";
+const ERC20_MAKER_PAYMENT: &str = "erc20MakerPayment";
+
+/// state index for `MakerPayment` structure from `EtomicSwapMakerV2.sol`
+///
+///     struct MakerPayment {
+///         bytes20 paymentHash;
+///         uint32 paymentLockTime;
+///         MakerPaymentState state;
+///     }
+const MAKER_PAYMENT_STATE_INDX: usize = 2;
+
+struct MakerPaymentArgs {
+    taker_address: Address,
+    taker_secret_hash: [u8; 32],
+    maker_secret_hash: [u8; 32],
+    payment_time_lock: u64,
+}
+
+struct MakerValidationArgs<'a> {
+    swap_id: Vec<u8>,
+    amount: U256,
+    taker: Address,
+    taker_secret_hash: &'a [u8],
+    maker_secret_hash: &'a [u8],
+    payment_time_lock: u64,
+}
+
+struct MakerRefundTimelockArgs {
+    payment_amount: U256,
+    taker_address: Address,
+    taker_secret_hash: [u8; 32],
+    maker_secret_hash: [u8; 32],
+    payment_time_lock: u64,
+    token_address: Address,
+}
```

Comment:
are these field size not known? 

```rus
taker_secret_hash: &'a [u8],
maker_secret_hash: &'a [u8],
```

Reply:
missed these fields. eth taker/maker secret hashes are 32 bytes, so we can do an additional validation using fixed size array, thanks
https://github.com/KomodoPlatform/komodo-defi-framework/pull/2211/commits/1e6ea8ee62775debaac27cf81772e3aa4625c106